File: .github/scripts/create-release.py
#!/usr/bin/env python3

import os
import sys
import requests
import subprocess

from glob import glob

token = os.getenv('GITHUB_API_TOKEN')
if not token:
    print("No API token found in environment", file=sys.stderr)
    exit(1)

release_tag = os.getenv('TAG_NAME')
if not release_tag:
    print("No tag found in environment", file=sys.stderr)
    exit(1)

print(f"# Current tag is {release_tag}")

base_headers = {
    'Accept': 'application/vnd.github.v3+json',
    'Authorization': f"token {token}",
}

url = 'https://api.github.com/repos/IBM/nodejs-idb-connector/tags'
headers = base_headers
r = requests.get(url)
r.raise_for_status()
tags = r.json()

# Find the tag prior to this one
for i, tag in enumerate(tags):
    if tag['name'] == release_tag:
        prior_tag = tags[i+1]
        break
del tag

commit = prior_tag['commit']['sha']

print(f"# Prior tag is {prior_tag['name']}")

args = [
  'git',
  'log',
  '--grep=Bump version',
  '--invert-grep',
  '--no-merges',
  '--no-decorate',
  '--pretty=format:- %s',
  f"{commit}.."
]

print("# Generating changelog")
r = subprocess.run(args, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
output = r.stdout.decode('utf-8').rstrip()

body = f"""# nodejs-idb-connector {release_tag}

## Changes

---

{output}
"""

url = 'https://api.github.com/repos/IBM/nodejs-idb-connector/releases'

payload = {
    'tag_name': release_tag,
    'name': release_tag,
    'body': body,
    'draft': True
}

headers = base_headers
print(f"# Creating release for {release_tag}")
r = requests.post(url, headers=headers, json=payload)
r.raise_for_status()

release = r.json()
print(f"# Release created: {release['html_url']}")

url = release['upload_url'].split('{')[0]

for path in glob('build/stage/*/db2ia-*.tar.gz'):
    name = path.split('/')[-1]
    print(f"# Uploading {path} as {name}")

    params = { 'name': name }
    extra_headers = { 'Content-Type': 'application/gzip' }
    headers = {**base_headers, **extra_headers}
    r = requests.post(url, headers=headers, params=params, data=open(name, 'rb'))
    r.raise_for_status()


print(f"# Publishing release")
url = release['url']
payload = { 'draft': True }
headers = base_headers
r = requests.patch(url, headers=headers, json=payload)
r.raise_for_status()
release = r.json()
print(f"# Release published at {release['html_url']}")

File: .github/create-release.js
const fs = require('fs');
const path = require('node:path');

const tag = process.env.TAG_NAME;
const token = process.env.GITHUB_TOKEN;

if (!tag) {
    console.error("Must set TAG_NAME environment variable");
    process.exit(1);
}
if (!token) {
    console.error("Must set GITHUB_TOKEN environment variable");
    process.exit(1);
}

const headers = {
    "Authorization": `Bearer ${token}`,
    "X-GitHub-Api-Version": "2022-11-28",
    "Accept": "application/vnd.github+json",
};

(async function script() {
    const url = 'https://api.github.com/repos/IBM/nodejs-idb-connector/releases';
    let response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify({
            tag_name: tag,
            generate_release_notes: true,
            draft: true,
        }),
        headers: headers,
    });

    if (!response.ok) {
        console.error(`ERROR: Getting release failed: ${response.status}`)
        console.error(`URL: ${tag_url}`);
        process.exit(1);
    }

    const release_obj = await response.json();

    try {
        // Returns a goofy URL "template", get rid of the parameters
        const upload_url = release_obj.upload_url.substr(0, release_obj.upload_url.indexOf('{'));;

        const file = "lib/db2ia.node";
        const name = "db2ia.node";
        console.log(`Uploading ${file} -> ${name}`);

        const file_upload_url = `${upload_url}?name=${name}`;
        response = await fetch(file_upload_url, {
            method: "POST",
            body: fs.readFileSync(file),
            headers: {
                "Content-Type": "application/octet-stream",
                ...headers,
            }
        });
    }
    catch (e) {
        console.error(e);
        console.error('Deleting release');
        const response = await fetch(release_obj.url, {
            method: 'DELETE',
            headers: headers,
        });

        if (!response.ok) {
            console.error(`Failed: ${response.status} ${response.statusText}`);
        }

        process.exit(1);
    }

    if (!response.ok) {
        console.error(`ERROR: Upload of ${file} failed: ${response.status}`);
        console.error(`URL: ${file_upload_url}`);

        console.error('Deleting release');
        const response = await fetch(release_obj.url, {
            method: 'DELETE',
            headers: headers,
        });

        if (!response.ok) {
            console.error(`Failed: ${response.status} ${response.statusText}`);
        }

        process.exit(1);
    }

    response = await fetch(release_obj.url, {
        method: 'PATCH',
        headers: headers,
        body: JSON.stringify({
            draft: false,
        }),
    });

    if (!response.ok) {
        console.error(`ERROR: Couldn't update release draft status`);
        console.error(`URL: ${release_obj.url}`);
        process.exit(1);
    }
})();
